src/decode/vec2d.rs

```diff
@@ -0,0 +1,74 @@
+use std::ops::{Index, IndexMut};
+
+/// A 2 dimensional matrix in row-major order backed by a contiguous `Vec`
+#[derive(Debug)]
+pub struct Vec2D<T> {
+    data: Vec<T>,
+    cols: usize,
+    rows: usize,
+}
+
+impl<T> Vec2D<T> {
+    /// Initialize a grid of size `rows * cols` with the given data element.
+    pub fn init(rows: usize, cols: usize, data: T) -> Vec2D<T>
```

Comment:
It'd be more logical to have the data as first parameter, to be consistent with the `vec![data; size]` style. Also, grouping together the dimensions as a tuple may make it more obvious that it's a size parameter:

`init(data: T, size: (usize, usize))`